src/server/master_server/chunk_server_manager.cc

```diff
@@ -1,69 +1,304 @@
 #include "src/server/master_server/chunk_server_manager.h"
 
+namespace protos {
+bool operator==(const protos::ChunkServerLocation& lhs,
+                const protos::ChunkServerLocation& rhs) {
+  return lhs.server_hostname() == rhs.server_hostname() &&
+         lhs.server_port() == rhs.server_port();
+}
+
+bool operator==(const protos::ChunkServer& lhs,
+                const protos::ChunkServer& rhs) {
+  return lhs.location() == rhs.location();
+}
+}  // namespace protos
+
 namespace gfs {
 
 namespace server {
 
-// Iterates through the chunk servers and selects the server with
-// the max available memory.
-// TODO(bmokutub): optimize this, maybe use a sorted map.
-std::shared_ptr<protos::ChunkServer> ChunkServerManager::AllocateChunkServer() {
-  auto servers_iterator = this->chunk_servers_map_.begin();
+ChunkServerLocationThreadSafeFlatSet ChunkServerManager::AllocateChunkServer(
+    const std::string& chunk_handle, const ushort& requested_servers_count) {
+  // Check if we have allocated servers for this previously
+  if (this->chunk_locations_map_.contains(chunk_handle)) {
+    // Chunk handle has been allocated previously.
+    // Return the previously allocated locations.
+    return this->chunk_locations_map_[chunk_handle];
+  }
+
+  ChunkServerLocationThreadSafeFlatSet allocated_locations;
```

Comment:
Furthermore, if you'd agree that we don't need a thread-safe structure, then a list (vector) would suffice here.

Reply:
Mentioned above